kernel/network/ethernet.c

```diff
@@ -0,0 +1,84 @@
+
+/******************************************************************************
+ *       ninjastorms - shuriken operating system                              *
+ *                                                                            *
+ *    Copyright (C) 2013 - 2016  Andreas Grapentin et al.                     *
+ *                                                                            *
+ *    This program is free software: you can redistribute it and/or modify    *
+ *    it under the terms of the GNU General Public License as published by    *
+ *    the Free Software Foundation, either version 3 of the License, or       *
+ *    (at your option) any later version.                                     *
+ *                                                                            *
+ *    This program is distributed in the hope that it will be useful,         *
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of          *
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
+ *    GNU General Public License for more details.                            *
+ *                                                                            *
+ *    You should have received a copy of the GNU General Public License       *
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.   *
+ ******************************************************************************/
+
+#include "ethernet.h"
+#include "kernel/network/arp.h"
+#include "kernel/network/e1000.h"
+#include "kernel/network/network_io.h"
+#include "kernel/logger/logger.h"
+
+#include <sys/types.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define INTTOHEXCHAR(val) ((val) > 9 ? (val)+'a'-10 : (val) + '0')
+
+void
+send_ethernet(mac_address_t dest_mac, ether_type eth_type, void *payload, size_t len_payload)
+{
+  // make sure ethernet frame is at least 60 bytes long
+  uint8_t len_padding = len_payload < 46 ? MINIMUM_PAYLOAD_LENGTH - len_payload : 0;
+  uint32_t len = len_padding + len_payload;
+  ethernet_frame_t *eth_frame = (ethernet_frame_t *) malloc(sizeof(ethernet_frame_t) + len);
+  eth_frame->dest_mac = hton_mac(dest_mac);
+  eth_frame->src_mac = hton_mac(my_mac());
+  eth_frame->ether_type = htons(eth_type);
+  memcpy(eth_frame->payload, payload, len);
+
+#ifdef ETHERNET_DEBUG
+  log_debug("Payload is %i, adding %i padding", len_payload, len_padding);
+#endif
+  if(len_padding != 0)
+    {
+      uint8_t padding[len_padding];
+      for(uint8_t i = 0; i < len_padding; i++)
+        padding[i] = 0x00;
```

Comment:
We cannot initialize variable-sized objects, so I filled the padding with a loop. Because the padding can be at maximum `MINIMUM_PAYLOAD_LENGTH` bytes long, we could initialize an array of this size and only copy `len_padding` bytes. Would that be preferable?